### PR TITLE
feat: Add sync APIs for "Deno.permissions"

### LINKED ISSUE #15576: suggestion: `Deno.permissions.querySync`
`Deno.permissions.query` is currently async. A sync version would be useful in contexts where top-level await would otherwise break node. Refs:
* denoland/deno_std#2446
* denoland/deno_std#1495
* denoland/deno_std#1396 (possibly)

In denoland/deno_std#1495 in particular, @bartlomieju mentions that a sync version of this API might not be a good idea. Is this still the case? If so, is there a decent workaround for the above-mentioned refs?

### PR DESCRIPTION
**Pull request by iuioiua**
This commit adds sync versions of async APIs to "Deno.permissions"
namespace.

Following APIs were added:
- "Deno.permissions.querySync"
- "Deno.permissions.requestSync"
- "Deno.permissions.revokeSync"

Closes #15576.

### CODE REVIEW

File: cli/tests/unit/permissions_test.ts [hunk 1]
```diff
@@ -13,12 +13,25 @@ Deno.test(async function permissionInvalidName() {
   }, TypeError);
 });
 
+Deno.test(function permissionInvalidNameSync() {
```
Comment (iuioiua) [inline cli/tests/unit/permissions_test.ts:16]: Separate tests have been created for `querySync()`. Should the test functions just be appened to the existing async tests or are they good as they are?

**Comment by bartlomieju** [comment]
If we are to add `querySync()` we should probably have `promptSync()` and `revokeSync()` too

**Comment by iuioiua** [comment]
I'm happy to do so in a couple follow up PRs if you're sure 🤙🏾

**Comment by bartlomieju** [comment]
> I'm happy to do so in a couple follow up PRs if you're sure 🤙🏾

Great, I think they should all be added in this PR

File: cli/tests/unit/permissions_test.ts (hunk 1 above)
Comment (bartlomieju) [reply cli/tests/unit/permissions_test.ts:16]: This seems fine

**Comment by iuioiua** [comment]
In the aim of minimising work, once `querySync()` looks good, I'll replicate the implementation for the other two.

File: cli/tsc/dts/lib.webworker.d.ts [hunk 2]
```diff
@@ -2777,6 +2777,7 @@ declare var PermissionStatus: {
 
 interface Permissions {
```
Comment (iuioiua) [inline cli/tsc/dts/lib.webworker.d.ts:2778]: Should this include `revoke()` and `request()`?

File: cli/tsc/dts/lib.dom.d.ts [hunk 3]
```diff
@@ -10909,6 +10909,7 @@ declare var PermissionStatus: {
 
 interface Permissions {
```
Comment (iuioiua) [inline cli/tsc/dts/lib.dom.d.ts:10910]: Should this include revoke() and request()?

**Comment by iuioiua** [comment]
I'm not sure why CI is failing. It may be unrelated to this change.

**Comment by iuioiua** [comment]
#17094 is a prerequisite for this PR.

File: cli/tsc/dts/lib.dom.d.ts (hunk 3 above)
Comment (iuioiua) [reply cli/tsc/dts/lib.dom.d.ts:10910]: @bartlomieju

Comment (bartlomieju) [reply cli/tsc/dts/lib.dom.d.ts:10910]: The changes to this file should be reverted - it's a file coming from TS and it describes the `DOM` library - these are browser permissions, not Deno permissions. The sync versions should only be added to `lib.deno.ns.d.ts`

**Review by kt3k — APPROVED** [review]
LGTM

**Comment by iuioiua** [comment]
*Fixing conflicts*

**Comment by iuioiua** [comment]
Conflicts fixed but a seemingly unrelated error has occurred.
```
/service-workers/cache-storage/cache-put.https.any.html


failures:

Cache.put called twice with request URLs that differ only by a fragment
assert_equals: Cache.put should replace existing response with new response. Attributes differ: statusText. expected "New status" but got "Old status"
    at http://web-platform.test:8000/service-workers/cache-storage/resources/test-helpers.js:194:9
    at Array.forEach (<anonymous>)
    at assert_response_equals (http://web-platform.test:8000/service-workers/cache-storage/resources/test-helpers.js:193:51)
    at http://web-platform.test:8000/service-workers/cache-storage/cache-put.https.any.js:223:11

failures:

        "Cache.put called twice with request URLs that differ only by a fragment"

file result: failed. 22 passed; 1 failed; 5 expected failure; total 28
```

**Comment by lino-levan** [comment]
LGTM

File: cli/tsc/dts/lib.deno.ns.d.ts [hunk 4]
```diff
@@ -512,6 +512,21 @@ declare namespace Deno {
...
+     *   name: "inherit",
+     *   permissions: {
+     *     net: "inherit",
+     *   },
+     *   fn() {
+     *     const status = Deno.permissions.querySync({ name: "net" })
+     *     assertEquals(status.state, "granted");
+     *   },
+     * });
+     * ```
```
Comment (dsherret) [inline cli/tsc/dts/lib.deno.ns.d.ts:529]: Duplicating all the examples and adding sync versions for these properties seems to add a bit too much noise. Maybe we should only document the async versions and remove these additions?

File: runtime/js/10_permissions.js [hunk 5]
```diff
@@ -197,6 +197,19 @@
       return PromiseResolve(cache(desc, state));
     }
 
+    querySync(desc) {
```
Comment (dsherret) [inline runtime/js/10_permissions.js:200]: These seem basically identical to the async versions. Perhaps we could update the async versions to just call the sync versions and have the result in a promise. That would help keep them in sync.

Comment (iuioiua) [reply runtime/js/10_permissions.js:200]: Good idea - will do.

Comment (iuioiua) [reply runtime/js/10_permissions.js:193]: Done

**Review by dsherret — APPROVED** [review]
LGTM. Thanks!

**Review by bartlomieju — APPROVED** [review]
LGTM to me too. Thanks @iuioiua, I'll update the commit message and land this.

**Pull request merged**